src/methods/join.js

```diff
@@ -807,18 +831,70 @@ export class JoinImplementation {
     }
   }
 
+  // Joins property bindings at a join point. Returns joined bindings and a set of leaked
+  // objects, based on the binding for the property "_isLeaked", which marks objects as
+  // leaked.
   joinPropertyBindings(
     realm: Realm,
     joinCondition: AbstractValue,
     m1: PropertyBindings,
     m2: PropertyBindings,
     c1: CreatedObjects,
-    c2: CreatedObjects
-  ): PropertyBindings {
-    let join = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+    c2: CreatedObjects,
+    g1: Generator,
+    g2: Generator
+  ): [PropertyBindings, IntermediateObjects] {
+    // Joins two property bindings of type "_isLeaked". If an object is leaked
+    // on one side, it is also leaked on the other side. Returns [x, y] where
+    // x is the joined descriptor, and y is set to the IntermediateObjectValue consisting of generator and object if the object leaked.
+    let joinLeakedObjectPropertyAndAdvise = (
+      b: PropertyBinding,
+      d1: void | Descriptor,
+      d2: void | Descriptor
+    ): [Descriptor | void, IntermediateObjectValue | void] => {
```

Comment:
Given our existing naming scheme, it is really confusing for an (ObjectValue, Generator) pair to be referred to as an IntermediateObjectValue. Every other type that ends with the Value suffix is a subclass of Value. This one is not. That is surprising and a trap for the unwary.